refactor timestamp parsing code to be less coupled with message parsing

bazsi opened the pull request:
Originally this branch attempted to do timezone estimation based on the incoming timestamp, however, that is going to be added in a subsequent PR.

This now only contains the refactor and a new unit test for timestamp parsing and should introduce no functional changes.

kira-syslogng commented:
Build SUCCESS

kira-syslogng commented:
Build SUCCESS

MrAnno commented:
Do we want to make this optional (`flags(no-guess-timezone)`, for example)? Do we need to turn this on by default?

When sending logs without timezone information from a file (a few hours from the past), some messages will have a timezone shift, others won't.  This could be noticed, but the 30-second accuracy trick makes it even harder for the user to figure out what's happening with their logs, that is why I'm asking about the conditionality and the default value of the heuristic.

faxm0dem commented:
IMHO this should be off by default for the reason given by @MrAnno

bazsi commented:
This pull request **fixes 1 alert** when merging 8c7135c7dbb1db340e185cbf8162d341eed29a82 into 82a7a012353143314d8482b7f249e56367a4da59 - [view on LGTM.com](https://lgtm.com/projects/g/balabit/syslog-ng/rev/pr-ccb38abc4030c8ea6b085fcc16700d15caac03cd)

**fixed alerts:**

* 1 for FIXME comment

---

*Comment posted by [LGTM.com](https://lgtm.com)*

bazsi commented:
This pull request **fixes 1 alert** when merging 0d54b901d30f1f7c66b5a16fe82c3c7f989458a1 into c20061c377a3be5eac33b443cdf9c71945e80723 - [view on LGTM.com](https://lgtm.com/projects/g/balabit/syslog-ng/rev/pr-c4693294683aeff5c4223456fd1d6684175fa1f4)

**fixed alerts:**

* 1 for FIXME comment

---

*Comment posted by [LGTM.com](https://lgtm.com)*

kira-syslogng commented:
Build FAILURE

kira-syslogng commented:
Build FAILURE

bazsi commented:
This pull request **fixes 1 alert** when merging d2d903c83037d5677b042af19e6d71754d0cdd69 into c20061c377a3be5eac33b443cdf9c71945e80723 - [view on LGTM.com](https://lgtm.com/projects/g/balabit/syslog-ng/rev/pr-412d022e5144bafee54f64c30c6e4000360743c9)

**fixed alerts:**

* 1 for FIXME comment

---

*Comment posted by [LGTM.com](https://lgtm.com)*

MrAnno commented on modules/syslogformat/syslog-timestamp.h:
```diff
@@ -0,0 +1,35 @@
+
+gboolean scan_rfc3164_timestamp(const guchar **data, gint *length, LogStamp *stamp,
+                                gboolean ignore_result, glong default_timezone);
+gboolean scan_rfc5424_timestamp(const guchar **data, gint *length, LogStamp *stamp,
```
They are beautiful. 😄

kira-syslogng commented:
Build FAILURE

bazsi commented:
This pull request **fixes 1 alert** when merging b0ba9ea3f7ff57498e3beffb439fec55d414ce0e into c20061c377a3be5eac33b443cdf9c71945e80723 - [view on LGTM.com](https://lgtm.com/projects/g/balabit/syslog-ng/rev/pr-9737510d27d3972104ab007143c92de21803343e)

**fixed alerts:**

* 1 for FIXME comment

---

*Comment posted by [LGTM.com](https://lgtm.com)*

bazsi commented:
This pull request **fixes 1 alert** when merging 5bddf12edac3d9416df115f7882a336f0aefdc82 into c20061c377a3be5eac33b443cdf9c71945e80723 - [view on LGTM.com](https://lgtm.com/projects/g/balabit/syslog-ng/rev/pr-57c613c26db977d705796645a3a2fcd6c4ace4f5)

**fixed alerts:**

* 1 for FIXME comment

---

*Comment posted by [LGTM.com](https://lgtm.com)*

kira-syslogng commented:
Build SUCCESS

kira-syslogng commented:
Build SUCCESS

bazsi commented:
This pull request **fixes 1 alert** when merging e3114ecebd3f94d10e8aec04dd710b9cda31e8c7 into 95748088aee53a07f1457a71ee92658452ad9655 - [view on LGTM.com](https://lgtm.com/projects/g/balabit/syslog-ng/rev/pr-170f5cd8818fdc7805f140a7a280571f3fb8883e)

**fixed alerts:**

* 1 for FIXME comment

---

*Comment posted by [LGTM.com](https://lgtm.com)*

kira-syslogng commented:
Build FAILURE

bazsi commented:
If this now becomes green, we are ready for a review.

kira-syslogng commented:
Build SUCCESS

bazsi commented:
This pull request **fixes 1 alert** when merging 7ac357d6845b3bc17e05fdb8acdd17fa66accfac into 95748088aee53a07f1457a71ee92658452ad9655 - [view on LGTM.com](https://lgtm.com/projects/g/balabit/syslog-ng/rev/pr-374b8247cd7f8ec900f01c1f505521ace2b41d10)

**fixed alerts:**

* 1 for FIXME comment

---

*Comment posted by [LGTM.com](https://lgtm.com)*

bazsi commented:
This pull request **fixes 1 alert** when merging 19579b8a650b9e1288d4c6b4d5558124b4f8a17c into 95748088aee53a07f1457a71ee92658452ad9655 - [view on LGTM.com](https://lgtm.com/projects/g/balabit/syslog-ng/rev/pr-e6e4d8b6e78b02c90b267b84ff7edb1860029137)

**fixed alerts:**

* 1 for FIXME comment

---

*Comment posted by [LGTM.com](https://lgtm.com)*

kira-syslogng commented:
Build FAILURE

bazsi commented:
This pull request **fixes 1 alert** when merging 19579b8a650b9e1288d4c6b4d5558124b4f8a17c into 95748088aee53a07f1457a71ee92658452ad9655 - [view on LGTM.com](https://lgtm.com/projects/g/balabit/syslog-ng/rev/pr-e6e4d8b6e78b02c90b267b84ff7edb1860029137)

**fixed alerts:**

* 1 for FIXME comment

---

*Comment posted by [LGTM.com](https://lgtm.com)*

bazsi commented:
This pull request **fixes 1 alert** when merging 6a86607e5edc18bf81fc6cb8b80795005d44a46b into 6c0307af0e0d3735bb104cc973f5b53e4c3dfa80 - [view on LGTM.com](https://lgtm.com/projects/g/balabit/syslog-ng/rev/pr-45d29658f164bd350ddd88b2843befd40b80731a)

**fixed alerts:**

* 1 for FIXME comment

---

*Comment posted by [LGTM.com](https://lgtm.com)*

bazsi commented:
This pull request **fixes 1 alert** when merging 5d6b553df84ed204e3f848c8ba076c1a1cbd63b5 into 1a2e4cd87c32322228090fa7b697f470325f07ec - [view on LGTM.com](https://lgtm.com/projects/g/balabit/syslog-ng/rev/pr-856c405d4d9a8529f0c3566780363dacd4f552a8)

**fixed alerts:**

* 1 for FIXME comment

---

*Comment posted by [LGTM.com](https://lgtm.com)*

kira-syslogng commented:
Build SUCCESS

Kokan commented on configure.ac:
```diff
@@ -467,6 +467,25 @@ if test "$no_pie_option_available" = "yes"; then
+
+
+if test $ostype = "Darwin"; then
+  LDFLAGS_FORCE_FLAT_NAMESPACE="-Wl,-force_flat_namespace"
```
I asked the same today. MacOS has a so called [two-level namespace linking](https://developer.apple.com/library/archive/documentation/Porting/Conceptual/PortingUnix/compiling/compiling.html#//apple_ref/doc/uid/TP40002850-BCIHJBBF) by default.
That prevents defining your own implementation for functions that are present in other libs. (like supplying your own `read` function).

That is a good feature by default, but for test it can prevent stubbing technics. The new test case uses it to fake time:
 https://github.com/balabit/syslog-ng/pull/2427/files#diff-e45df9d43f2cc8ab23500d13ac59e5e2R32

bazsi replied:
As @Kokan explained, it allows a unit test to inject function implementations defined in other shared libraries. the timestamp parsing test uses it to fake the current time by providing a cached_g_current_time() function.

https://www.unix.com/man-page/osx/1/ld/

The default behavior of the linker on macOS differs from what Linux does, this switch restores that behavior.

bazsi commented on modules/syslogformat/syslog-format.c:
```diff
@@ -572,11 +572,6 @@ log_msg_parse_rfc5424_date_unnormalized(LogMessage *self, const guchar **data, g
-error:
-  /* no recognizable timestamp, use current time */
-
-  self->timestamps[LM_TS_STAMP] = self->timestamps[LM_TS_RECVD];
```
If you look at the caller of this function, it does the same assignment in case this function returns FALSE, so we basically assigned the same value twice.

bazsi commented:
Hmm.. my force_single_namespace trick does not work on macOS anymore, as I've moved the code from being in a module (modules/syslogformat) to the core itself.

I'll still need to experiment with how this could be done on macOS, without macro trickery. :(

bazsi commented:
I have now reworked the time mocking code so it should be more portable again. Now, I am basically 
using the cache variable to store the faked time and added a function to timeutils to set that variable.

I hope it this remains green this way.

Otherwise the branch should not have material changes since the last review round.

bazsi commented:
This pull request **fixes 1 alert** when merging 65ba45f564201c5719560a33e9d0165ad4128fc3 into b45361150605a85358e80b24e4679311ffc291dc - [view on LGTM.com](https://lgtm.com/projects/g/balabit/syslog-ng/rev/pr-c8d1f76e0c97918f79fa07da90ef7c3c997959af)

**fixed alerts:**

* 1 for FIXME comment

---

*Comment posted by [LGTM.com](https://lgtm.com)*

kira-syslogng commented:
Build SUCCESS

bazsi commented:
It's green \o/

Kokan commented on modules/syslogformat/tests/Makefile.am:
```diff
@@ -0,0 +1,13 @@
+
+modules_syslogformat_tests_test_parse_timestamp_SOURCES	= modules/syslogformat/tests/test_parse_timestamp.c
+modules_syslogformat_tests_test_parse_timestamp_CFLAGS	= $(TEST_CFLAGS) -I$(top_srcdir)/modules/syslogformat
+modules_syslogformat_tests_test_parse_timestamp_LDFLAGS	= $(LDFLAGS_FORCE_FLAT_NAMESPACE)
```
Have you considered adding `force-flat-namespace` for all of the unit tests just like in `cmake` version ?

bazsi commented:
I have removed this at the end completely.

On Thu, Dec 20, 2018, 21:57 Kókai Péter <notifications@github.com wrote:

> *@Kokan* commented on this pull request.
> ------------------------------
>
> In modules/syslogformat/tests/Makefile.am
> <https://github.com/balabit/syslog-ng/pull/2427#discussion_r243415265>:
>
> > @@ -0,0 +1,13 @@
> +modules_syslogformat_tests_TESTS		= \
> +	modules/syslogformat/tests/test_parse_timestamp
> +
> +check_PROGRAMS				+= ${modules_syslogformat_tests_TESTS}
> +
> +modules_syslogformat_tests_test_parse_timestamp_SOURCES	= modules/syslogformat/tests/test_parse_timestamp.c
> +modules_syslogformat_tests_test_parse_timestamp_CFLAGS	= $(TEST_CFLAGS) -I$(top_srcdir)/modules/syslogformat
> +modules_syslogformat_tests_test_parse_timestamp_LDFLAGS	= $(LDFLAGS_FORCE_FLAT_NAMESPACE)
>
> Have you considered adding force-flat-namespace for all of the unit tests
> just like in cmake version ?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/balabit/syslog-ng/pull/2427#pullrequestreview-187181523>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AArldj6wQx4T5vcn38AoKGQ7Yz3C4xw_ks5u6_nGgaJpZM4YxpZm>
> .
>

bazsi commented on modules/syslogformat/syslog-format.c:
```diff
@@ -1037,6 +754,12 @@ log_msg_parse_legacy(const MsgFormatOptions *parse_options,
+
+  if (!log_msg_extract_cisco_timestamp_attributes(self, &src, &left, parse_options->flags))
+    {
+      goto error;
```
Good point. I've modified the call-site so that it ignores return value, just like with log_msg_parse_seq(), which also has an ignored return value.

The last two patch I've just pushed should unify the conventions here.

bazsi commented on lib/timeutils.c:
```diff
@@ -103,6 +103,9 @@ TLS_BLOCK_START
 TLS_BLOCK_END;
 
+/* this indicates that a test program is faking the current time */
+gboolean faking_time;
```
The mocking didn't work out using linker trickery at  the end. Whereas -force_flat_namespace did work in the original location (when the function to be mocked was in libsyslog-ng.so, and the call-site was in libsyslog-format.so). This broke when I moved the stuff into the core, lib/str-timestamp directory.

Probably, when linking libsyslog-ng.so, the internal call-sites to published functions do not go through ld.so's symbol resolution, rather a simple call instruction is used. At that point I gave up the linker based tricks.

I could make the "faking_time" variable thread specific, but that will not improve much. I made an effort not to have to modify the source code for the sake of tests, but at the end I couldn't. This is still a "bit" better than if I'd have to compile the core with a preprocessor define and use macros to override stuff, at least I think.

With all this said, I couldn't find an easy way to mock internal symbols, so I just added support for the testability cases. It's not that bad :)

Let me know if you disagree, then we can find a better solution.

MrAnno replied:
Thank you for the detailed explanation. :+1: 
`faking_time` should be good enough in this case.

bazsi commented on lib/timeutils.h:
```diff
@@ -37,6 +37,7 @@ void clean_time_cache(void);
 
 
 void invalidate_cached_time(void);
+void set_cached_time(GTimeVal *timeval);
```
See my previous comments. I could add a comment there.

MrAnno replied:
I'd add a comment here, just to avoid confusion, but it's not a must for me.

bazsi commented:
This pull request **fixes 1 alert** when merging 100d202e1548cbc1dd74c6b3d5884e3c00cb40a3 into 426ec2b8c73112f13dd4c95a9161aed0e6931ce8 - [view on LGTM.com](https://lgtm.com/projects/g/balabit/syslog-ng/rev/pr-9469b0f729a7fdebec35c3963b23e3f4e6dbb594)

**fixed alerts:**

* 1 for FIXME comment

---

*Comment posted by [LGTM.com](https://lgtm.com)*

CaronB1962 commented:
Unsubscribes immediately 

Caron.bradley@yahoo.com

> On Dec 25, 2018, at 12:36 AM, László Szemere <notifications@github.com> wrote:
> 
> Merged #2427 into master.
> 
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub, or mute the thread.

MrAnno commented:
@CaronB1962 You can unsubscribe by clicking on the "unwatch" button here:
https://github.com/balabit/syslog-ng
